### ISSUE TITLE
Improve parsing of Type2 charstrings with |callsubr| and |callgsubr| operators

### ISSUE THREAD

**Issue by Snuffleupagus** [mozilla/pdf.js#5929]
Our current [Type2 charstring parser](https://github.com/mozilla/pdf.js/blob/master/src/core/fonts.js#L6173) is sometimes rejecting valid Type1C font data, with a message such as e.g. `"Warning: Not enough parameters for rlineto; actual: 0, expected: 2"`.

To address this, it appears that we need to improve the Type2 charstring parsing to handle the `callsubr` and `callgsubr` operators in order to obtain correct stack data.

___

**Examples of PDF files affected by this:**
 - Bug [849481](https://bugzilla.mozilla.org/show_bug.cgi?id=849481) - https://bug849481.bugzilla.mozilla.org/attachment.cgi?id=8423217
 - Issue #3694 - https://i.cloudup.com/oG1S4DIPAa.pdf.
 - Issue #3117 - https://www.dropbox.com/s/09eg9pj6pf4g0ir/hk-2020.pdf
 - Issue #3001 - http://www.ccs.neu.edu/home/dvanhorn/talks/umd-trustworthy-2013.pdf#page=18

**Relevant specifications:**
 - http://partners.adobe.com/public/developer/en/font/5176.CFF.pdf
 - http://partners.adobe.com/public/developer/en/font/5177.Type2.pdf

**Excerpt from IRC:**
```
16:15	Snuffleupagus	So, I'm wondering if we're not always clearing/resetting the stack properly.
16:15	Snuffleupagus	Setting |undefStack = true| after line https://github.com/mozilla/pdf.js/blob/master/src/core/fonts.js#L6275 seem to fix a handful of font issues (and it passes tests on Windows). Is this simply a stupid idea, or would it make sense to do that?
16:22	yury	Snuffleupagus: you might be on to something
16:23	Snuffleupagus	I'll submit a PR, to get testing done on Linux too!
16:26	yury	Snuffleupagus: actually stack is defined there
16:26	yury	it's empty
16:27	yury	Snuffleupagus: making it undefStack = true will just disable the check
16:27	Snuffleupagus	Yeah, thus why I said that it might be a stupid idea :)
16:27	yury	are we doing wrong parsing or something?
16:28	yury	it's actually a wrong thing we are doing in https://github.com/mozilla/pdf.js/blob/master/src/core/fonts.js#L6246
16:29	* Snuffleupagus is trying to read the Type2 spec
16:29	yury	we let pass maybe-invalid strings
16:30	yury	Snuffleupagus: we need parsing of the callsubr, callgsubr to set a proper stack
16:30	yury	and remove undefStack variable
16:31	Snuffleupagus	That sounds "fun" to implement xD
16:31	yury	Snuffleupagus: looks like medium size task, can you open a bug and list PDFs there and paste the log ^
16:32	Snuffleupagus	Sure, I'll do that later tonight!
16:33	yury	shall not be that hard, just probably we parse charstring as we receive them, instead of saving them and allowing callsubr to lookup the callee
```

**Comment by timvandermeij** [comment]
Perhaps #3694 is also an example of such an issue?

**Comment by fbender** [comment]
@Snuffleupagus Any idea when this issue can be resolved?

**Comment by yurydelendik** [comment]
@fbender is there are reason why you are posting to issues with such rate? pinging people will not speed up the issue resolution.

**Comment by fbender** [comment]
Sorry, I'm kind of triaging forms related issues atm. Stumbled over a couple of issues (not form-related as well) that seemed abandoned (sometimes with code ready to be merged) and wanted to ping some people to figure out the current status, issue relationships or get estimates. I'm pretty much done now, so no more notifications from me unless something is happening to issues or dependencies.

**Comment by yurydelendik** [comment]
@fbender if you are interested in an issue, please add a reason why you are interested in it, e.g. "my web site has PDFs affected by this issue and I have 1 billion visits per day (see http://example.org/my-page.html)". This might help to elevate its priority and better get attention on the contributor and reviewers.

**Comment by fbender** [comment]
OK, I will try to do that next time. For now, I only looked at issues that I have experienced at one point or affects a sizable population (e.g. this bug affects some PDFs of non-English regions, e.g. the German speaking area where Fx has a sizable installation base -- so this is mostly a QA / product polish issue and should receive a high priority, especially considering deprecation of NPAPI plugins), or issues blocking these two kinds of issues.
